compiler/packages/babel-plugin-react-compiler/src/Entrypoint/Imports.ts

```diff
@@ -42,50 +53,175 @@ export function validateRestrictedImports(
   }
 }
 
-export function addImportsToProgram(
-  path: NodePath<t.Program>,
-  importList: Array<ExternalFunction>,
-): void {
-  const identifiers: Set<string> = new Set();
-  const sortedImports: Map<string, Array<string>> = new Map();
-  for (const {importSpecifierName, source} of importList) {
-    /*
-     * Codegen currently does not rename import specifiers, so we do additional
-     * validation here
+export class ProgramContext {
+  uids: Set<string> = new Set();
+  hookPattern: string | null;
+  scope: BabelScope;
+  imports: Map<string, Map<string, NonLocalImportSpecifier>> = new Map();
+  reactRuntimeModule: string;
+  constructor(
+    program: NodePath<t.Program>,
+    reactRuntimeModule: CompilerReactTarget,
+    hookPattern: string | null,
+  ) {
+    this.hookPattern = hookPattern;
+    this.scope = program.scope;
+    this.reactRuntimeModule = getReactCompilerRuntimeModule(reactRuntimeModule);
+  }
+  isHookName(name: string): boolean {
+    if (this.hookPattern == null) {
+      return isHookName(name);
+    } else {
+      const match = new RegExp(this.hookPattern).exec(name);
+      return (
+        match != null && typeof match[1] === 'string' && isHookName(match[1])
+      );
+    }
+  }
+  hasReference(name: string): boolean {
+    return (
+      this.uids.has(name) ||
+      this.scope.hasBinding(name) ||
+      this.scope.hasGlobal(name) ||
+      this.scope.hasReference(name)
```

Comment:
this also prevents names that might be shadowed in local scopes, right?

Reply:
Yep, I took this from Babel's implementation (https://github.com/babel/babel/blob/main/packages/babel-traverse/src/scope/index.ts#L510-L513).